File: ds-algorithms-specialization/fibonacci-huge/fibonacci_huge.cpp
#include <iostream>
#include <vector>

using namespace std;

long long get_fibonacci_huge_naive(long long n, long long m) {
    if (n <= 1)
        return n;

    long long previous = 0;
    long long current  = 1;

    for (long long i = 0; i < n - 1; ++i) {
        long long tmp_previous = previous;
        previous = current;
        current = tmp_previous + current;
    }

    return current % m;
}

bool is_a_period(vector<long long> &numbers) {
    int size = numbers.size();
    int half_size = size / 2;
    bool is_a_period = size > 0 && size % 2 == 0;
    for (int i = 0; is_a_period && i < half_size; ++i) {
        is_a_period = numbers[i] == numbers[i + half_size];
    }
    return is_a_period;
}

vector<long long> find_period(long long m) {
    vector<long long> period;

    long long previous = 0 % m;
    long long current  = 1 % m;

    period.push_back(previous);
    period.push_back(current);

    while (true) {
        long long tmp_previous = previous;
        previous = current;
        current = (tmp_previous + current) % m;
        period.push_back(current);
        if (is_a_period(period)) {
            period.erase (period.begin() + (period.size() / 2), period.end());
            break;
        }
    }
    return period;
}

long long get_fibonacci_huge_fast(long long n, long long m) {
    if (n <= 1)
        return n;

    vector<long long> period = find_period(m);
    int size = period.size();

    return period[n % size];
}

int main() {
    long long n, m;
    std::cin >> n >> m;
    std::cout << get_fibonacci_huge_fast(n, m) << '\n';
}

File: ds-algorithms-specialization/fibonacci-huge/fibonacci_huge.py
# Uses python3
import sys


def get_fibonacci_huge_naive(n, m):
    if n <= 1:
        return n

    previous = 0
    current  = 1

    for _ in range(n - 1):
        previous, current = current, previous + current

    return current % m


def is_a_period(numbers):
    size = len(numbers)
    half = int(size / 2)
    if size > 0 and size % 2 == 0:
        for i in range(half):
            if numbers[i] != numbers[i + half]:
                return False
    else:
        return False
    return True


def find_period(m):
    previous = 0
    current  = 1

    period = [previous % m, current % m]

    while True:
        previous, current = current, (previous + current) % m
        period.append(current)
        if is_a_period(period):
            period = period[int(len(period)/2):]
            break
    return period


def get_fibonacci_huge_fast(n, m):
    if n <= 1:
        return n

    period = find_period(m)
    size = len(period)

    return period[n % size]


if __name__ == '__main__':
    input = sys.stdin.read()
    n, m = map(int, input.split())
    print(get_fibonacci_huge_fast(n, m))
